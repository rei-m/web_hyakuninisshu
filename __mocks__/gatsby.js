const React = require("react");
const { action } = require("@storybook/addon-actions");
const iconImage = require("@src/images/app-icon.png");
const correctImage = require("@src/images/check_correct.png");
const incorrectImage = require("@src/images/check_incorrect.png");
const dogezaImage = require("@src/images/dogeza_businessman.png");
const karutaImage = require("@src/images/karuta_001.jpg");
const tatamiImage  = require("@src/images/tatami_part.png");

const linkActionHandler = action("Link:");
const navigateActionHandler = action("NavigateTo:");

const imageSrc = (image) => (image.default === "test-file-stub" ? image.default : image);

module.exports = {
  graphql: (args) => args,
  Link: ({ to, ...rest }) =>
    React.createElement("a", {
      ...rest,
      href: to,
      onClick: (e) => {
        e.preventDefault();
        linkActionHandler(to);
      }
    }),
  StaticQuery: () => {},
  useStaticQuery: (args) => {
    const query = args[0];
    if (query.indexOf("query SEOQuery") > -1) {
      return {
        site: {
          siteMetadata: {
            title: "百人一首 簡単に暗記",
            description: "説明",
            author: "rei-m",
          },
        },
        ogpImage: {
          publicURL: imageSrc(iconImage),
        },
      };
    }
    if (query.indexOf("query CorrectImageQuery") > -1) {
      return {
        correctImage: {
          childImageSharp: {
            fluid: {
              aspectRatio: 0.7228915662650602,
              srcSet: "",
              sizes: "(max-width: 300px) 100vw, 200px",
              src: imageSrc(correctImage),
            },
          },
        },
        incorrectImage: {
          childImageSharp: {
            fluid: {
              aspectRatio: 0.7228915662650602,
              srcSet: "",
              sizes: "(max-width: 300px) 100vw, 200px",
              src: imageSrc(incorrectImage),
            },
          },
        },
      };
    }
    if (query.indexOf("query DogezaImageQuery") > -1) {
      return {
        dogezaImage: {
          childImageSharp: {
            fluid: {
              aspectRatio: 0.7228915662650602,
              srcSet: "",
              sizes: "(max-width: 200px) 200px, 200px",
              src: imageSrc(dogezaImage),
            },
          },
        },
      };
    }
    if (query.indexOf("query TatamiImageQuery") > -1) {
      return {
        tatamiImage: {
          publicURL: tatamiImage,
        },
      };
    }
    if (query.indexOf("query KarutaImageQuery") > -1) {
      return {
        karutaImages: {
          edges: Array.from(Array(100).keys()).map(i => ({
            node: {
              childImageSharp: {
                fluid: {
                  aspectRatio: 0.7228915662650602,
                  srcSet: "",
                  sizes: "(max-width: 200px) 100vw, 200px",
                  src: imageSrc(karutaImage),
                }
              },
              name: `karuta_${("00" + (i + 1)).slice(-3)}`
            }
          }))
        }
      };
    }
    return {};
  },
  navigate: (to, options) => {
    navigateActionHandler(to, options);
  }
}
