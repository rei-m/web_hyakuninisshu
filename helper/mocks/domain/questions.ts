import { Question } from '@src/domain/models';

export const MOCK_QUESTION_1: Question = {
  id: 1,
  correctAnswerKarutaNo: 1,
  choiceKarutaNoList: [1, 2, 3, 4],
};

export const MOCK_QUESTION_2: Question = {
  id: 2,
  correctAnswerKarutaNo: 2,
  choiceKarutaNoList: [1, 2, 3, 4],
};

export const MOCK_QUESTION_1_STARTED: Question = {
  id: 1,
  correctAnswerKarutaNo: 1,
  choiceKarutaNoList: [1, 2, 3, 4],
  startTime: 1583595557624,
};

export const MOCK_QUESTION_1_ANSWERED_CORRECT: Question = {
  id: 1,
  correctAnswerKarutaNo: 1,
  choiceKarutaNoList: [1, 2, 3, 4],
  startTime: 1583595557624,
  answer: {
    isCorrect: true,
    answerTime: 1583595559624,
  },
};

export const MOCK_QUESTION_2_ANSWERED_WRONG: Question = {
  id: 2,
  correctAnswerKarutaNo: 2,
  choiceKarutaNoList: [1, 2, 3, 4],
  startTime: 1583595689123,
  answer: {
    isCorrect: false,
    answerTime: 1583595690123,
  },
};
