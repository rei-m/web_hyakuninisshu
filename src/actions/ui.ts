import { Action, Dispatch } from 'redux';
import { Color, Kimariji } from '@src/types';

export const OPEN_KARUTAS_FILTER_NAME = 'OPEN_KARUTAS_FILTER';
export type OPEN_KARUTAS_FILTER_TYPE = typeof OPEN_KARUTAS_FILTER_NAME;

export const CLOSE_KARUTAS_FILTER_NAME = 'CLOSE_KARUTAS_FILTER';
export type CLOSE_KARUTAS_FILTER_TYPE = typeof CLOSE_KARUTAS_FILTER_NAME;

export const TOGGLE_KARUTAS_COLOR_NAME = 'TOGGLE_KARUTAS_COLOR';
export type TOGGLE_KARUTAS_COLOR_TYPE = typeof TOGGLE_KARUTAS_COLOR_NAME;

export const TOGGLE_KARUTAS_KIMARIJI_NAME = 'TOGGLE_KARUTAS_KIMARIJI_NAME';
export type TOGGLE_KARUTAS_KIMARIJI_TYPE = typeof TOGGLE_KARUTAS_KIMARIJI_NAME;

export interface OpenKarutasFilterAction extends Action {
  readonly type: OPEN_KARUTAS_FILTER_TYPE;
}

export interface CloseKarutasFilterAction extends Action {
  readonly type: CLOSE_KARUTAS_FILTER_TYPE;
}

export interface OKarutasFilterAction extends Action {
  readonly type: OPEN_KARUTAS_FILTER_TYPE;
}

export interface ToggleKarutasColorAction extends Action {
  readonly type: TOGGLE_KARUTAS_COLOR_TYPE;
  readonly payload: {
    readonly color: Color;
    readonly checked: boolean;
  };
}

export interface ToggleKarutasKimarijiAction extends Action {
  readonly type: TOGGLE_KARUTAS_KIMARIJI_TYPE;
  readonly payload: {
    readonly kimariji: Kimariji;
    readonly checked: boolean;
  };
}

export type UiActions =
  | OpenKarutasFilterAction
  | CloseKarutasFilterAction
  | ToggleKarutasColorAction
  | ToggleKarutasKimarijiAction;

/*
 * action creators
 */
export const openKarutasFilter = () => (dispatch: Dispatch<OpenKarutasFilterAction>) =>
  dispatch({
    type: OPEN_KARUTAS_FILTER_NAME,
  });

export const closeKarutasFilter = () => (dispatch: Dispatch<CloseKarutasFilterAction>) =>
  dispatch({
    type: CLOSE_KARUTAS_FILTER_NAME,
  });

export const toggleKarutasColor = (color: Color, checked: boolean) => (
  dispatch: Dispatch<ToggleKarutasColorAction>
) => {
  const action: ToggleKarutasColorAction = {
    payload: {
      color,
      checked,
    },
    type: TOGGLE_KARUTAS_COLOR_NAME,
  };

  dispatch(action);
};

export const toggleKarutasKimariji = (kimariji: Kimariji, checked: boolean) => (
  dispatch: Dispatch<ToggleKarutasKimarijiAction>
) => {
  const action: ToggleKarutasKimarijiAction = {
    payload: {
      kimariji,
      checked,
    },
    type: TOGGLE_KARUTAS_KIMARIJI_NAME,
  };

  dispatch(action);
};
