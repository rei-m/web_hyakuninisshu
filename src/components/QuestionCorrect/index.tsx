import * as React from 'react';
import styled from 'styled-components';
import { withState } from 'recompose';
import { Dialog } from '@blueprintjs/core';
import { withAppTheme } from '@src/styles';
import { Karuta } from '@src/types';
import Tatami from '@src/components/Tatami';
import KarutaCard from '@src/components/KarutaCard';
import { toKarutaIdString, toKimarijiString } from '@src/components/helper';

export interface QuestionCorrectProps {
  readonly karuta: Karuta;
  readonly isAllAnswered: boolean;
  readonly onClickGoToNext: () => void;
  readonly onClickGoToResult: () => void;
}

const Frame = Tatami.extend`
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  min-height: calc(100vh - ${({ theme }) => theme.headerHeight});

  @media screen and (min-width: ${({ theme }) => theme.minWidthWide}) {
    min-height: calc(100vh - ${({ theme }) => theme.headerHeightWide});
  }
`;

const Header = styled.div`
  position: relative;
  width: 160px;
`;

const NumberAndKimariji = withAppTheme(styled.div)`
  padding: ${({ theme }) => `${theme.spacing1x} ${theme.spacing2x}`};
  border: 1px solid #d3d3d3;
  width: 160px;
  margin: 0 auto;
  font-size: 1.2rem;
  background-color: #ffffff;
`;

const OpenDetail = styled.button`
  position: absolute;
  top: 0;
  right: -48px;
  height: 33px;
  width: 33px;
`;

const KarutaFrame = withAppTheme(styled.div)`
  width: 200px;
  height: 260px;
  border: 6px solid ${({ theme }) => theme.colorPrimaryDark};
  border-radius: 10px;
  display: flex;
  align-items: center;
  justify-content: center;
  margin: ${({ theme }) => theme.spacing2x} auto 0 auto;
  background-color: ${({ theme }) => theme.colorThin};
  box-shadow: 0 -1px 5px rgba(0, 0, 0, 0.26);
  font-family: 'Sawarabi Mincho';
`;

const Inner = styled.div`
  display: flex;
  flex-direction: row-reverse;
  justify-content: center;
`;

const Phrase = styled.div`
  width: 1.9rem;
  font-size: 1.9rem;
  line-height: 2rem;
`;

const SecondPhrase = Phrase.extend`
  padding-top: 24px;
  margin-right: 8px;
`;

const ThirdPhrase = Phrase.extend`
  padding-top: 48px;
  margin-right: 8px;
`;

const FourthPhrase = Phrase.extend`
  padding-top: 32px;
  margin-right: 8px;
`;

const FifthPhrase = Phrase.extend`
  padding-top: 56px;
  margin-right: 8px;
`;

const Creator = styled.div`
  width: 1.6rem;
  margin-right: 16px;
  font-size: 1.4rem;
  line-height: 1.6rem;
  align-self: flex-end;
`;

const NextButton = styled.button`
  margin-top: 32px;
`;

const enhance = withState<{}, boolean, 'opened', 'setOpened'>(
  'opened',
  'setOpened',
  false
);

const QuestionCorrect = enhance<
  QuestionCorrectProps & {
    opened: boolean;
    setOpened: (state: boolean) => boolean;
  }
>(
  ({
    karuta,
    isAllAnswered,
    onClickGoToNext,
    onClickGoToResult,
    opened,
    setOpened
  }) => {
    const onClickOpenDetail = () => {
      setOpened(true);
    };

    const onCloseDialog = () => {
      setOpened(false);
    };

    return (
      <Frame>
        <Header>
          <NumberAndKimariji>
            {toKarutaIdString(karuta.id)} / {toKimarijiString(karuta.kimariji)}
          </NumberAndKimariji>
          <OpenDetail
            onClick={onClickOpenDetail}
            className="pt-button"
            data-test="open-detail"
          >
            詳
          </OpenDetail>
        </Header>
        <KarutaFrame>
          <Inner>
            <Phrase>{karuta.firstKanji}</Phrase>
            <SecondPhrase>{karuta.secondKanji}</SecondPhrase>
            <ThirdPhrase>{karuta.thirdKanji}</ThirdPhrase>
            <FourthPhrase>{karuta.fourthKanji}</FourthPhrase>
            <FifthPhrase>{karuta.fifthKanji}</FifthPhrase>
            <Creator>{karuta.creator}</Creator>
          </Inner>
        </KarutaFrame>
        {isAllAnswered ? (
          <NextButton
            onClick={onClickGoToResult}
            className="pt-button pt-large pt-icon-double-chevron-right"
            data-test="go-to-result"
          >
            結果を見る
          </NextButton>
        ) : (
          <NextButton
            onClick={onClickGoToNext}
            className="pt-button pt-large pt-icon-double-chevron-right"
            data-test="go-to-next"
          >
            次へ進む
          </NextButton>
        )}
        <Dialog
          isOpen={opened}
          onClose={onCloseDialog}
          title="正解"
          style={{
            maxWidth: 380,
            padding: 0,
            width: '80vw'
          }}
        >
          <KarutaCard karuta={karuta} />
        </Dialog>
      </Frame>
    );
  }
);

export default QuestionCorrect;
