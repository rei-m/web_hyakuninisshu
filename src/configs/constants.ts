export const APP_TITLE = '百人一首 - 簡単に暗記 -';
