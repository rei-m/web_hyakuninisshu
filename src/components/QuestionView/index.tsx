import * as React from 'react';
import { graphql, StaticQuery } from 'gatsby';
import styled from '@src/styles/styled-components';
import YomiFudaView from '@src/components/YomiFudaView';
import ToriFudaView from '@src/components/ToriFudaView';
import QuestionResult from '@src/components/QuestionResult';
import { Answer, Question, ToriFuda } from '@src/types';

export interface Props {
  question: Question;
  answer?: Answer;
  totalCount: number;
  currentPosition: number;
  onClickToriFuda: (toriFuda: ToriFuda) => void;
  onClickResult: () => void;
}

interface QueryData {
  questionBGImage: {
    publicURL: string;
  };
}

const Container = styled.div<{ bgImageUrl: string }>`
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  background-image: url("${({ bgImageUrl }) => bgImageUrl}");
  min-height: calc(100vh - ${({ theme }) => theme.headerHeight});
  @media screen and (min-width: ${({ theme }) => theme.minWidthWide}) {
    min-height: calc(100vh - ${({ theme }) => theme.headerHeightWide});
  }
`;

const YomiFudaBox = styled.div`
  position: relative;
  margin-bottom: ${({ theme }) => theme.spacing3x};
  width: 312px;
`;

const ToriFudaBox = styled.div`
  display: flex;
  justify-content: center;
  flex-direction: row-reverse;
`;

const Position = styled.div`
  position: absolute;
  right: 0;
  top: -32px;
  color: #fff;
  border-bottom: 1px dotted #fff;
`;

const QuestionView: React.FC<Props> = ({
  answer,
  currentPosition,
  question,
  totalCount,
  onClickToriFuda,
  onClickResult,
}) => (
  <StaticQuery
    query={query}
    render={({ questionBGImage }: QueryData) => (
      <Container bgImageUrl={questionBGImage.publicURL}>
        <YomiFudaBox>
          <Position>
            {currentPosition} / {totalCount}
          </Position>
          <YomiFudaView yomiFuda={question.yomiFuda} style={{ margin: 'auto' }} />
        </YomiFudaBox>
        <ToriFudaBox>
          {question.toriFudas.map((toriFuda, i) => (
            <ToriFudaView
              toriFuda={toriFuda}
              style={{
                margin: '8px',
                opacity: !answer || answer.karutaNo === toriFuda.karutaNo ? 1 : 0.8,
              }}
              key={i}
              onClick={onClickToriFuda}
              data-test={`torifuda-${i}`}
            />
          ))}
        </ToriFudaBox>
        {answer && <QuestionResult answer={answer} onClick={onClickResult} data-test="result" />}
      </Container>
    )}
  />
);

export default QuestionView;

const query = graphql`
  query {
    questionBGImage: file(relativePath: { eq: "tatami_part.png" }) {
      publicURL
    }
  }
`;
