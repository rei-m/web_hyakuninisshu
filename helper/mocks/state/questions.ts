import { questionsTypes } from '@src/state/questions';

export const MOCK_YOMIFUDA_1: questionsTypes.YomiFuda = {
  karutaNo: 1,
  shoku: '秋の田の',
  niku: 'かりほの庵の',
  sanku: '苫をあらみ',
};

export const MOCK_TORIFUDA_1: questionsTypes.ToriFuda = {
  karutaNo: 1,
  shiku: 'きりにぬれつつ',
  kekku: 'わがころもでは',
};

export const MOCK_TORIFUDA_2: questionsTypes.ToriFuda = {
  karutaNo: 2,
  shiku: 'ころもほすてふ',
  kekku: 'あまのかぐやま',
};

export const MOCK_TORIFUDA_3: questionsTypes.ToriFuda = {
  karutaNo: 3,
  shiku: 'ながながしよを',
  kekku: 'ひとりかもねむ',
};

export const MOCK_TORIFUDA_4: questionsTypes.ToriFuda = {
  karutaNo: 4,
  shiku: 'ふじのたかねに',
  kekku: 'ゆきはふりつつ',
};
