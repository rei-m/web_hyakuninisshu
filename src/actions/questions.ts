import { Action, Dispatch } from 'redux';
import { GlobalState } from '@src/state';
import { Answer, Karuta, Question } from '@src/types';
import { randomizeArray } from '@src/utils';
import { fetchTorifudas, questionsFilter, toDulation } from '@src/utils/questions';
import {
  ColorCondition,
  KarutaStyleCondition,
  KimarijiCondition,
  QuestionAnimCondition,
  QuestionState,
  RangeFromCondition,
  RangeToCondition,
} from '@src/enums';

export const START_TRAINING_NAME = 'START_TRAINING_NAME';
export type START_TRAINING_TYPE = typeof START_TRAINING_NAME;

export const START_EXAM_NAME = 'START_EXAM_NAME';
export type START_EXAM_TYPE = typeof START_EXAM_NAME;

export const RESTART_QUESTIONS_NAME = 'RESTART_QUESTIONS_NAME';
export type RESTART_QUESTIONS_TYPE = typeof RESTART_QUESTIONS_NAME;

export const ANSWER_QUESTION_NAME = 'ANSWER_QUESTION_NAME';
export type ANSWER_QUESTION_TYPE = typeof ANSWER_QUESTION_NAME;

export const CONFIRM_CORRECT_NAME = 'CONFIRM_CORRECT_NAME';
export type CONFIRM_CORRECT_TYPE = typeof CONFIRM_CORRECT_NAME;

export const OPEN_NEXT_QUESTION_NAME = 'OPEN_NEXT_QUESTION_NAME';
export type OPEN_NEXT_QUESTION_TYPE = typeof OPEN_NEXT_QUESTION_NAME;

export const FINISH_QUESTIONS_NAME = 'FINISH_QUESTIONS_NAME';
export type FINISH_QUESTIONS_TYPE = typeof FINISH_QUESTIONS_NAME;

export interface StartTrainingAction extends Action {
  readonly type: START_TRAINING_TYPE;
  readonly payload: {
    readonly karutas: Karuta[];
    readonly questions: Question[];
    readonly startedTime: number;
    readonly nextState: QuestionState;
    readonly dulation: number;
  };
  readonly meta: {
    readonly rangeFrom: RangeFromCondition;
    readonly rangeTo: RangeToCondition;
    readonly kimariji: KimarijiCondition;
    readonly color: ColorCondition;
    readonly kamiNoKuStyle: KarutaStyleCondition;
    readonly shimoNoKuStyle: KarutaStyleCondition;
    readonly questionAnim: QuestionAnimCondition;
  };
}

export interface StartExamAction extends Action {
  readonly type: START_EXAM_TYPE;
  readonly payload: {
    readonly karutas: Karuta[];
    readonly questions: Question[];
    readonly startedTime: number;
    readonly nextState: QuestionState;
  };
}

export interface RestartQuestionsAction extends Action {
  readonly type: RESTART_QUESTIONS_TYPE;
  readonly payload: {
    readonly questions: Question[];
    readonly startedTime: number;
    readonly nextState: QuestionState;
  };
}

export interface AnswerQuestionAction extends Action {
  readonly type: ANSWER_QUESTION_TYPE;
  readonly payload: {
    readonly answer: Answer;
    readonly nextState: QuestionState;
  };
}

export interface ConfirmCorrectAction extends Action {
  readonly type: CONFIRM_CORRECT_TYPE;
  readonly payload: {
    readonly nextState: QuestionState;
  };
}

export interface OpenNextQuestionAction extends Action {
  readonly type: OPEN_NEXT_QUESTION_TYPE;
  readonly payload: {
    readonly nextIndex: number;
    readonly nextState: QuestionState;
    readonly startedTime: number;
  };
}

export interface FinishQuestionsAction extends Action {
  readonly type: FINISH_QUESTIONS_TYPE;
  readonly payload: {
    readonly nextState: QuestionState;
  };
}

export type QuestionsActions =
  | StartTrainingAction
  | StartExamAction
  | RestartQuestionsAction
  | AnswerQuestionAction
  | ConfirmCorrectAction
  | OpenNextQuestionAction
  | FinishQuestionsAction;

/*
 * action creators
 */
export const startTraining = (
  karutas: Karuta[],
  rangeFrom: RangeFromCondition,
  rangeTo: RangeToCondition,
  kimariji: KimarijiCondition,
  color: ColorCondition,
  kamiNoKuStyle: KarutaStyleCondition,
  shimoNoKuStyle: KarutaStyleCondition,
  questionAnim: QuestionAnimCondition
) => (dispatch: Dispatch<QuestionsActions>) => {
  const questions = new QuestionsFactory(karutas)
    .setRange(rangeFrom, rangeTo)
    .setKimariji(kimariji)
    .setColor(color)
    .setKamiNoKuStyle(kamiNoKuStyle)
    .setShimoNoKuStyle(shimoNoKuStyle)
    .create();

  const action: StartTrainingAction = {
    meta: {
      color,
      kamiNoKuStyle,
      kimariji,
      rangeFrom,
      rangeTo,
      shimoNoKuStyle,
      questionAnim,
    },
    payload: {
      karutas,
      nextState: QuestionState.InAnswer,
      questions: randomizeArray<Question>(questions),
      startedTime: new Date().getTime(),
      dulation: toDulation(questionAnim),
    },
    type: START_TRAINING_NAME,
  };

  dispatch(action);
};

export const startExam = (karutas: Karuta[]) => (dispatch: Dispatch<QuestionsActions>) => {
  const questions = new QuestionsFactory(karutas)
    .setRange(RangeFromCondition.One, RangeToCondition.OneHundred)
    .setKimariji(KimarijiCondition.None)
    .setColor(ColorCondition.None)
    .setKamiNoKuStyle(KarutaStyleCondition.KanjiAndKana)
    .setShimoNoKuStyle(KarutaStyleCondition.KanaOnly)
    .create();

  const action: StartExamAction = {
    payload: {
      karutas,
      nextState: QuestionState.InAnswer,
      questions: randomizeArray(questions),
      startedTime: new Date().getTime(),
    },
    type: START_EXAM_NAME,
  };

  dispatch(action);
};

export const restartQuestions = () => (dispatch: Dispatch<QuestionsActions>, getState: () => GlobalState) => {
  const { questions, answers } = getState().questions;
  // TODO: 全て回答済みでなかったらエラー
  const finder: { [questionId: number]: Question } = questions.reduce((previous, current) => {
    return { ...previous, [current.id]: current };
  }, {});
  const targets = answers
    .filter(a => !a.correct)
    .map(a => finder[a.questionId])
    .map(q => ({ ...q, toriFudas: randomizeArray(q.toriFudas) }));

  const action: RestartQuestionsAction = {
    payload: {
      nextState: QuestionState.InAnswer,
      questions: randomizeArray(targets),
      startedTime: new Date().getTime(),
    },
    type: RESTART_QUESTIONS_NAME,
  };

  dispatch(action);
};

export const answerQuestion = (questionId: number, karutaNo: number) => (
  dispatch: Dispatch<QuestionsActions>,
  getState: () => GlobalState
) => {
  const { questions, lastStartedTime } = getState().questions;
  const question = questions.find(q => q.id === questionId)!;
  // TODO: QuestionとlastStartedTimeがundefinedの場合
  const correct = question.correctKaruta.no === karutaNo;
  const time = new Date().getTime() - lastStartedTime!;
  const answer = {
    correct,
    karutaNo,
    questionId,
    time,
  };
  const action: AnswerQuestionAction = {
    payload: {
      answer,
      nextState: QuestionState.Answered,
    },
    type: ANSWER_QUESTION_NAME,
  };

  dispatch(action);
};

export const confirmCorrect = (): ConfirmCorrectAction => {
  return {
    payload: {
      nextState: QuestionState.ConfirmCorrect,
    },
    type: CONFIRM_CORRECT_NAME,
  };
};

export const openNextQuestion = () => (dispatch: Dispatch<QuestionsActions>, getState: () => GlobalState) => {
  const { currentIndex } = getState().questions;
  const nextIndex = currentIndex + 1;
  const action: OpenNextQuestionAction = {
    payload: {
      nextIndex,
      nextState: QuestionState.InAnswer,
      startedTime: new Date().getTime(),
    },
    type: OPEN_NEXT_QUESTION_NAME,
  };
  dispatch(action);
};

export const finishQuestions = (): FinishQuestionsAction => {
  return {
    payload: {
      nextState: QuestionState.Finished,
    },
    type: FINISH_QUESTIONS_NAME,
  };
};

class QuestionsFactory {
  private karutas: Karuta[];
  private rangeFrom: number;
  private rangeTo: number;
  private kimariji: number;
  private color: string;
  private kamiNoKuStyle: number;
  private shimoNoKuStyle: number;

  constructor(karutas: Karuta[]) {
    this.karutas = karutas;
    this.rangeFrom = RangeFromCondition.One;
    this.rangeTo = RangeToCondition.OneHundred;
    this.kimariji = KimarijiCondition.None;
    this.color = ColorCondition.None;
    this.kamiNoKuStyle = KarutaStyleCondition.KanjiAndKana;
    this.shimoNoKuStyle = KarutaStyleCondition.KanaOnly;
  }

  public setRange(rangeFrom: number, rangeTo: number) {
    this.rangeFrom = rangeFrom;
    this.rangeTo = rangeTo;
    return this;
  }

  public setKimariji(kimariji: number) {
    this.kimariji = kimariji;
    return this;
  }

  public setColor(color: string) {
    this.color = color;
    return this;
  }

  public setKamiNoKuStyle(kamiNoKuStyle: number) {
    this.kamiNoKuStyle = kamiNoKuStyle;
    return this;
  }

  public setShimoNoKuStyle(shimoNoKuStyle: number) {
    this.shimoNoKuStyle = shimoNoKuStyle;
    return this;
  }

  public create() {
    const targetKarutas = questionsFilter(this.karutas)(this.rangeFrom, this.rangeTo)(this.kimariji)(this.color);

    return targetKarutas.map((k, i) => {
      const id = i + 1;
      const correctKaruta = k;
      const yomiFuda =
        this.kamiNoKuStyle === 0
          ? {
              firstText: k.firstKanji,
              karutaNo: k.no,
              questionId: id,
              secondText: k.secondKanji,
              thirdText: k.thirdKanji,
            }
          : {
              firstText: k.firstKana,
              karutaNo: k.no,
              questionId: id,
              secondText: k.secondKana,
              thirdText: k.thirdKana,
            };
      const toriFudas = fetchTorifudas(this.karutas, correctKaruta).map(toriFuda => {
        return this.shimoNoKuStyle === 0
          ? {
              fifthText: toriFuda.fifthKanji,
              fourthText: toriFuda.fourthKanji,
              karutaNo: toriFuda.no,
              questionId: id,
            }
          : {
              fifthText: toriFuda.fifthKana,
              fourthText: toriFuda.fourthKana,
              karutaNo: toriFuda.no,
              questionId: id,
            };
      });
      return { id, correctKaruta, yomiFuda, toriFudas };
    });
  }
}
