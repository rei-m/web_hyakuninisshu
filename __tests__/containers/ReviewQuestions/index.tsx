import { GlobalState } from '@src/state';
import * as React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import { navigate } from 'gatsby';
import {
  mapDispatchToProps,
  mapStateToProps,
  OwnProps,
  Props,
  ReviewQuestions,
} from '@src/containers/organisms/ReviewQuestions';
import ReviewInitializer from '@src/containers/ReviewInitializer';
import QuestionView from '@src/components/organisms/KarutaPlaying';
import QuestionCorrect from '@src/components/organisms/KarutaPlayingCorrect';
import { MockStore } from 'redux-mock-store';
import { initialState as questionsState } from '@src/state/questions/reducers';
import { initialState as uiState } from '@src/state/ui/reducers';
import { ROUTE_PATHS } from '@src/constants';
import { QuestionState } from '@src/enums';
import { Answer, Question, ToriFuda, YomiFuda } from '@src/types';
import { create } from '@test/factories';
import { mockAppStoreCreateor } from '@test/helpers';
import { ANSWER_QUESTION_NAME, CONFIRM_CORRECT_NAME, OPEN_NEXT_QUESTION_NAME } from '@src/state/questions/constants';

const createQuestion = (id: number) => {
  const yomiFuda = create<YomiFuda>('yomiFuda', {
    questionId: id,
  });

  const toriFudas = [...Array(4).keys()].map(_ =>
    create<ToriFuda>('toriFuda', {
      questionId: id,
    })
  );

  return create<Question>('question', {
    id,
    toriFudas,
    yomiFuda,
  });
};

describe('<ReviewQuestions />', () => {
  describe('components', () => {
    let wrapper: ShallowWrapper;
    let baseProps: Props;

    beforeEach(() => {
      const submitTime = 1000;

      baseProps = {
        submitTime,
        totalCount: 0,
        currentPosition: 0,
        dulation: 0.6,
        onClickToriFuda: jest.fn(),
        onClickResult: jest.fn(),
        onClickGoToNext: jest.fn(),
        onClickGoToResult: jest.fn(),
      };
    });

    describe('when question is not started', () => {
      it('should render ReviewInitializer', () => {
        wrapper = shallow(<ReviewQuestions {...baseProps} />);
        expect(wrapper.find(ReviewInitializer).length).toBe(1);
      });
    });

    describe('when in answer', () => {
      it('should render QuestionView', () => {
        const props = {
          ...baseProps,
          lastStartedTime: 11000,
          questionState: QuestionState.InAnswer,
          question: createQuestion(1),
        };
        wrapper = shallow(<ReviewQuestions {...props} />);
        expect(wrapper.find(QuestionView).length).toBe(1);
      });
    });

    describe('when after answered', () => {
      it('should render QuestionCorrect', () => {
        const props = {
          ...baseProps,
          lastStartedTime: 11000,
          questionState: QuestionState.ConfirmCorrect,
          question: createQuestion(1),
        };
        wrapper = shallow(<ReviewQuestions {...props} />);
        expect(wrapper.find(QuestionCorrect).length).toBe(1);
      });
    });

    describe('when finished', () => {
      it('should render QuestionCorrect', () => {
        const props = {
          ...baseProps,
          lastStartedTime: 11000,
          questionState: QuestionState.Finished,
          question: createQuestion(1),
        };
        wrapper = shallow(<ReviewQuestions {...props} />);
        expect(wrapper.find(QuestionCorrect).length).toBe(1);
      });
    });
  });

  describe('mapStateToProps', () => {
    it('should convert state to props', () => {
      const submitTime = 1000;
      const props: OwnProps = {
        submitTime,
      };

      const state: GlobalState = {
        questions: {
          ...questionsState,
          answers: [create<Answer>('answer')],
          lastStartedTime: 11000,
          questionState: QuestionState.Finished,
          questions: [createQuestion(1)],
        },
        ui: uiState,
      };

      expect(mapStateToProps(state, props)).toMatchSnapshot();
    });
  });

  describe('mapDispatchToProps', () => {
    let mockStore: MockStore<GlobalState>;

    beforeEach(() => {
      mockStore = mockAppStoreCreateor()({
        questions: { ...questionsState, questions: [createQuestion(1)] },
        ui: uiState,
      });
    });

    it('should dispatch openNextQuestion action when onClickGoToNext fired', () => {
      mapDispatchToProps(mockStore.dispatch).onClickGoToNext();
      const mockActions = mockStore.getActions();
      expect(mockActions[0].type).toEqual(OPEN_NEXT_QUESTION_NAME);
    });

    it('should dispatch confirmCorrect action when onClickResult fired', () => {
      mapDispatchToProps(mockStore.dispatch).onClickResult();
      const mockActions = mockStore.getActions();
      expect(mockActions[0].type).toEqual(CONFIRM_CORRECT_NAME);
    });

    it('should dispatch answerQuestion action when onClickToriFuda fired', () => {
      mapDispatchToProps(mockStore.dispatch).onClickToriFuda(mockStore.getState().questions.questions[0].toriFudas[0]);
      const mockActions = mockStore.getActions();
      expect(mockActions[0].type).toEqual(ANSWER_QUESTION_NAME);
    });

    it('should navigate to result when onClickGoToResult fired', () => {
      mapDispatchToProps(mockStore.dispatch).onClickGoToResult();
      expect(navigate).toHaveBeenCalledWith(ROUTE_PATHS.TRAINING_RESULT, { replace: true });
    });
  });
});
