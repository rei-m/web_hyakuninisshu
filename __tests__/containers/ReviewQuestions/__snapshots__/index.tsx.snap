// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ReviewQuestions /> mapStateToProps should convert state to props 1`] = `
Object {
  "answer": Object {
    "correct": true,
    "karutaNo": 1,
    "questionId": 1,
    "time": 1000000,
  },
  "currentPosition": 1,
  "dulation": 0,
  "lastStartedTime": 11000,
  "question": Object {
    "correctKaruta": Object {
      "color": "blue",
      "creator": "詠み手",
      "fifthKana": "ごく",
      "fifthKanji": "五句",
      "firstKana": "しょく",
      "firstKanji": "初句",
      "fourthKana": "よんく",
      "fourthKanji": "四句",
      "id": 4,
      "imageNo": "001",
      "kimariji": 1,
      "no": 1,
      "secondKana": "にく",
      "secondKanji": "二句",
      "thirdKana": "さんく",
      "thirdKanji": "三句",
      "translation": "歌の訳",
    },
    "id": 1,
    "toriFudas": Array [
      Object {
        "fifthText": "ごく",
        "fourthText": "よんく",
        "karutaNo": 1,
        "questionId": 1,
      },
      Object {
        "fifthText": "ごく",
        "fourthText": "よんく",
        "karutaNo": 1,
        "questionId": 1,
      },
      Object {
        "fifthText": "ごく",
        "fourthText": "よんく",
        "karutaNo": 1,
        "questionId": 1,
      },
      Object {
        "fifthText": "ごく",
        "fourthText": "よんく",
        "karutaNo": 1,
        "questionId": 1,
      },
    ],
    "yomiFuda": Object {
      "firstText": "初句",
      "karutaNo": 1,
      "questionId": 1,
      "secondText": "二句",
      "thirdText": "三句",
    },
  },
  "questionState": 3,
  "submitTime": 1000,
  "totalCount": 1,
}
`;
